### ISSUE TITLE
Create better hierarchy for network handling in vAPP and VM

### ISSUE THREAD

**Issue by vbauzysvmware** [terraform-providers/terraform-provider-vcd#264]
Ref: https://www.terraform.io/docs/providers/vcd/r/vapp_vm.html

Currently even in our example we have:
```
resource "vcd_vapp" "web" {
  name = "web"

  depends_on = ["vcd_network_direct.net"]
}
```
Which means that vAPP depends on network which VM uses. It's kind'a tricky to understand cause this only needed for destroying resource. As network can't be deleted until vAPP deleted cause network was attached.

Proposal is to change vAPP and VM network handling in that way:
```
resource "vcd_vapp" "web" {
  name = "web"
  networks =  ["vcd_network_direct.net", "net-my1, "vapp_net1"] // maybe better name attached_networks or similar
 depends_on = ["net-my1, "vapp_net1"]
}
```
And in VM change code behaviour that it won't automatically attach network to vAPP. Just complain and fail. In this way we would explicitly state that vAPP handles attach/detach vAPP networks and VM just uses what is available from vAPP. And we remove issue with strange implicit depending of network on vAPP through VM.

P.s. In UI we also have network management through vAPP, not VM. In VM we just assign available networks from vAPP networks set.

**Comment by vbauzysvmware** [comment]
cc: @lvirbalas

**Comment by Didainius** [comment]
Do we need a github tag for 3.0? I guess we agreed that this is correct behavior but this will break things.

**Comment by vbauzysvmware** [comment]
On strategy how make still back compatible and if it possible is another question. And when to do this change. I think for a moment, you could outline breaking things, cause you have most knowledge at the moment. I register this issue, to not forget what I figure out some time ago and see opportunity for very good improvement.

**Comment by lvirbalas** [comment]
> ~Do we need a github tag for 3.0?~ (I see breaking-change one) I guess we agreed that this is correct behavior but this will break things.

Let's agree to add "breaking-change" label for all non-backwards compatible issues that we could address with v3.0.0. Right now this is the only one with such label. CC @dataclouder

**Comment by lvirbalas** [comment]
Practical scenario which this refactoring would fix:
1. Create two Org networks.
2. Attach both to VM of some vApp.
3. Change config to leave only the first one attached to VM. Apply.

Expected result: **vApp** has only one network attached.
Result: **vApp** still has both networks attached, because we can't remove them when removing from VM as other VMs may need it.

**Comment by vbauzysvmware** [comment]
https://github.com/terraform-providers/terraform-provider-vcd/pull/455 solves this issue by providing new resources vcd_vapp_org_network and vcd_org_network. As results VM resource now can reference vapp networks directly. E.g.
```
resource "vcd_vapp_vm" "web1" { 
...
  network {
    type             = "org"
    name           = vcd_vapp_org_network.netResourceName.org_network_name
    ip                 = "10.10.104.161"
    ip_allocation_mode = "MANUAL"
    is_primary         = true
  }
}
```

**Comment by lvirbalas** [comment]
With #455 not only you can reference vApp networks in VM's `network` block, but also you gain the ability to define/attach these networks correctly at the vApp level:

```
# 2.7.0
resource "vcd_vapp_org_network" "demo_vapp_net_org" {
  vapp_name         = vcd_vapp.demo_vapp.name
  org_network_name  = vcd_network_routed.demo_routed_net_web.name
}

resource "vcd_vapp_vm" "demo_vm_wordpress" {
  vapp_name    = "${vcd_vapp.demo_vapp.name}"
  name         = "demo-vm-wordpress"
  catalog_name = "shared-catalog"
  template_name = "bitnami-wordpress-5.2.2-3-linux-centos-7-x86_64"
  memory        = 512
  cpus = 1

  network_dhcp_wait_seconds = 300
  network {
    type               = "org"
    name = vcd_vapp_org_network.demo_vapp_net_org.org_network_name
    ip_allocation_mode = "DHCP"
    is_primary         = true
  }

  accept_all_eulas = "true"
}
```

**Comment by vbauzysvmware** [comment]
Solved in #455 and will be in next release.
